### ISSUE TITLE
[$80] Remove individual calls for creating milestones for a timeline

### ISSUE THREAD

**Issue by RishiRajSahu** [appirio-tech/connect-app#2227]
Right now we are calling milestone's POST call multiple times to create milestones for a timeline. Now we have backend supported with implicit creation of milestone when calling POST for milestone with `templatedId` param passed, so this single call will create timeline and all the associated milestone using value of `templatedId`

**Comment by narekcat** [comment]
@gondzo @RishiRajSahu I start connect-app locally by running `npm start`, but when I try to create new app, it return HTTP error 500. In development console it says HTTP error 404 when requests to this URL  https://api.topcoder-dev.com/v4/projects/6704/phases by method GET. Are there any known issues with development back end ?

**Comment by gondzo** [comment]
Which account are you testing with?
available accounts are posted in the challenge forums

**Comment by narekcat** [comment]
@gondzo Ok, I have started to work on this issue before you posted available accounts in the form and create my own (username: testcat, email: test-cat@mail.ru). But ok no problem I'll try with accounts posted in the forum. Thank you.

**Comment by narekcat** [comment]
@gondzo @RishiRajSahu The description is ambiguous, I can't understand. How many endpoints we have in back end now ? One for creating timeline and one for creating milestones ? Or only one endpoint for creating timeline and all milestones together ?

**Comment by narekcat** [comment]
@gondzo @RishiRajSahu The description is ambiguous, I can't understand. How many endpoints we have in back end now ? One for creating timeline and one for creating milestones ? Or only one endpoint for creating timeline and all milestones together ? And if endpoint is only one, which one is /timelines or /milestones ?

**Comment by gondzo** [comment]
@narekcat create timeline endpoint will accept templateId parameter and will create all the necessary milestones from the templates in the backend. You just need to pass the templateId (productTemplateId) and and remove calls to create individual milestones. 
Here is the relevant backend change - https://github.com/topcoder-platform/tc-project-service/commit/827b2252947eb2e3a18f5732ebc09dc7d3187963#diff-7a3be78f0970e178e8ace2b5c70b17baR43

**Comment by narekcat** [comment]
@gondzo Yes I already understand it from repository https://github.com/topcoder-platform/tc-project-service/tree/feature/timeline-milestone, but there are some issue, because I did as you say, but milestones didn't created.

**Comment by gondzo** [comment]
Can you post api call details here (url, body) so we can investigate further

**Comment by narekcat** [comment]
@gondzo Api call details are:
Request URL: https://api.topcoder-dev.com/v4/timelines
Request Method: POST
Status Code: 201 
Remote Address: 34.231.208.179:443
Referrer Policy: no-referrer-when-downgrade

Request payload:
`{"param":{"name":"Welcome to the App phase","description":"This is the first stage in our project. We’re going to show you the detailed plan in your timeline, with all the milestones.","startDate":"2018-07-24T20:00:00+00:00","endDate":"2018-07-25T16:34:39+00:00","reference":"product","referenceId":765,"templateId":5}}`

Response:

{"id":"d6db479f-660c-460a-ba74-6de610b9f32d","version":"v4","result":{"success":true,"status":201,"content":{"createdAt":"2018-07-25T16:34:41.000Z","updatedAt":"2018-07-25T16:34:41.000Z","id":80,"name":"Welcome to the App phase","description":"This is the first stage in our project. We’re going to show you the detailed plan in your timeline, with all the milestones.","startDate":"2018-07-24T20:00:00.000Z","endDate":"2018-07-25T16:34:39.000Z","reference":"product","referenceId":765,"createdBy":40152856,"updatedBy":40152856,"milestones":[]},"metadata":{"totalCount":1}}}

**Comment by gondzo** [comment]
Product id 5 doesn't have any milestones defined. You can check defined milestones by calling
https://api.topcoder-dev.com/v4/productTemplates/5/milestones

productId=1 or productId=15 have some templates (website or visual design)

**Comment by narekcat** [comment]
Yes, really it returns milestones for productId=1:
`{"id":"16041685-195c-4af5-a54f-c6417485915f","version":"v4","result":{"success":true,"status":201,"content":{"createdAt":"2018-07-25T17:19:45.000Z","updatedAt":"2018-07-25T17:19:45.000Z","id":83,"name":"Welcome to the Website Development phase","description":"This is the first stage in our project. We’re going to show you the detailed plan in your timeline, with all the milestones.","startDate":"2018-07-24T20:00:00.000Z","endDate":"2018-08-18T17:19:43.000Z","reference":"product","referenceId":769,"createdBy":40152856,"updatedBy":40152856,"milestones":[{"id":263,"name":"community-work","description":"Please review and answer all the questions on the specification document before we can proceed","duration":5,"startDate":"2018-07-24T20:00:00.000Z","endDate":"2018-07-28T20:00:00.000Z","status":"reviewed","type":"community-work","details":{},"order":2,"plannedText":"dummy plannedText","activeText":"dummy activeText","completedText":"dummy completedText","blockedText":"dummy blockedText","hidden":false,"createdAt":"2018-07-25T17:19:45.000Z","updatedAt":"2018-07-25T17:19:45.000Z","createdBy":40152856,"updatedBy":40152856},{"id":264,"name":"phase-specification","description":"Please review and answer all the questions on the specification document before we can proceed","duration":5,"startDate":"2018-07-29T20:00:00.000Z","endDate":"2018-08-02T20:00:00.000Z","status":"reviewed","type":"phase-specification","details":{},"order":1,"plannedText":"dummy plannedText","activeText":"dummy activeText","completedText":"dummy completedText","blockedText":"dummy blockedText","hidden":false,"createdAt":"2018-07-25T17:19:45.000Z","updatedAt":"2018-07-25T17:19:45.000Z","createdBy":40152856,"updatedBy":40152856},{"id":265,"name":"community-review","description":"Please review and answer all the questions on the specification document before we can proceed","duration":5,"startDate":"2018-08-03T20:00:00.000Z","endDate":"2018-08-07T20:00:00.000Z","status":"reviewed","type":"community-review","details":{},"order":3,"plannedText":"dummy plannedText","activeText":"dummy activeText","completedText":"dummy completedText","blockedText":"dummy blockedText","hidden":false,"createdAt":"2018-07-25T17:19:45.000Z","updatedAt":"2018-07-25T17:19:45.000Z","createdBy":40152856,"updatedBy":40152856},{"id":266,"name":"delivery","description":"Please review and answer all the questions on the specification document before we can proceed","duration":5,"startDate":"2018-08-08T20:00:00.000Z","endDate":"2018-08-12T20:00:00.000Z","status":"reviewed","type":"delivery","details":{},"order":5,"plannedText":"dummy plannedText","activeText":"dummy activeText","completedText":"dummy completedText","blockedText":"dummy blockedText","hidden":false,"createdAt":"2018-07-25T17:19:45.000Z","updatedAt":"2018-07-25T17:19:45.000Z","createdBy":40152856,"updatedBy":40152856},{"id":267,"name":"final-fix","description":"Please review and answer all the questions on the specification document before we can proceed","duration":5,"startDate":"2018-08-13T20:00:00.000Z","endDate":"2018-08-17T20:00:00.000Z","status":"reviewed","type":"final-fix","details":{},"order":4,"plannedText":"dummy plannedText","activeText":"dummy activeText","completedText":"dummy completedText","blockedText":"dummy blockedText","hidden":true,"createdAt":"2018-07-25T17:19:45.000Z","updatedAt":"2018-07-25T17:19:45.000Z","createdBy":40152856,"updatedBy":40152856}]},"metadata":{"totalCount":1}}}`
But it doesn't shows milestones under Tileline tab, even after reloading.

**Comment by narekcat** [comment]
@gondzo I think I find what is the problem. In getting all milestoneTemplates, it sorts templates by order column(https://github.com/topcoder-platform/tc-project-service/blob/feature/timeline-milestone/src/routes/milestoneTemplates/list.js#L43), but when I call /timelines with templateId param, it just find all milestoneTemplates without sording(https://github.com/topcoder-platform/tc-project-service/blob/feature/timeline-milestone/src/routes/timelines/create.js#L61). Actually it saves without sorting and returns timeline data without sorting.

**Comment by gondzo** [comment]
I'm not sure if that's the only cause. 
Anyhow, can you create a PR with the changes for connect-app, and we can debug project service in a separate issue

FYI @vikasrohit

**Comment by narekcat** [comment]
Ok, I agree with you. The problem is with back end, because for timeline created in old flow shows, but for timeline created in new flow doesn't show. In both cases create response says ok and returns correct data.

**Comment by RishiRajSahu** [comment]
@narekcat Right now we have milestone templates for Website and Visual-Designs only So use these projects only for now to continue dev work for timeline-milestones. @gondzo could you add milestone templates for more projects and keep note of that maybe in code documents on the challenge.

**Comment by narekcat** [comment]
@RishiRajSahu I use Website for timeline-milestones, but there are issue with back end, if I'm right.

**Comment by RishiRajSahu** [comment]
@narekcat what issue is it causing if you are getting milestone templates without sorting ? anyway as per the scope of this issue your PR looks good, you can address my comment and I will merge the PR and close this issue. Also, for the backend issue which you want to address, could be taken care in new ticket here - https://github.com/topcoder-platform/tc-project-service/issues , feel free to create new issue if you think its valid.

**Comment by narekcat** [comment]
@RishiRajSahu Ok, thank you, I'll examine back end again.

**Comment by RishiRajSahu** [comment]
@narekcat you can address my comments on PR.

**Comment by narekcat** [comment]
@RishiRajSahu Oh, sorry, I have put all your comments on PR.
